### PR TITLE
Add Hailo inference backend

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

- add lazy HailoRT AutoBackend inference for Ultralytics-exported HEF directories
- convert YOLOv8/YOLO11 on-device NMS output and decode YOLO26 branch-first one-to-one outputs
- reuse the existing BaseBackend, metadata, anchor, box-decoding, and end-to-end prediction contracts
- replace the Hailo docs limitation with direct `predict` and `val` guidance

## Validation

- `pytest -q tests/test_python.py --disable-warnings --maxfail=1` — 109 passed, 2 skipped
- Ruff, Prettier, reference generation, dispatch checks, and synthetic NMS/YOLO26 decoder checks
- HailoRT 4.23 and 5.3 expose the same blocking VStreams API used here
- Hailo-8L hardware validation requested from @JESUSROYETH

Deleted: the obsolete no-inference warning and manual YOLO26 decoder instructions; no existing runtime backend could be reused because Hailo was the sole export format without one.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

🚀 Adds native HailoRT inference support for Ultralytics Hailo exports, including automatic YOLOv8, YOLO11, and YOLO26 output decoding.

### 📊 Key Changes

- Added a new `HailoBackend` for loading Hailo `.hef` files and running inference through HailoRT.
- Integrated Hailo support into `AutoBackend` for models in `*_hailo_model/` export directories.
- Added automatic post-processing for:
  - YOLOv8 and YOLO11 HailoRT NMS outputs.
  - YOLO26 raw, NMS-free outputs.
- Added validation to ensure Hailo inference is used only with detection models.
- Added documentation and API reference pages for the Hailo backend.
- Updated the Hailo integration guide with direct `predict` usage and metadata requirements.
- Bumped the Ultralytics version to `8.4.99`.

### 🎯 Purpose & Impact

- ✅ Enables direct `predict` and `val` workflows on Hailo hardware through Ultralytics and HailoRT.
- ⚡ Simplifies deployment by automatically locating the HEF file and decoding model-specific outputs.
- 🧩 Preserves compatibility with TAPPAS, GStreamer, and Raspberry Pi Hailo helper pipelines for custom applications.
- 📦 Requires HailoRT on the target device and the exported `metadata.yaml` to remain alongside the HEF file.
- 📚 Improves developer experience with dedicated backend documentation and clearer troubleshooting guidance.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated review message is here to help with the integration process, and an engineer will assist if anything needs attention. Please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/autobackend.py [hunk 1]
```diff
@@ -160,6 +162,7 @@ class AutoBackend(nn.Module):
         "deepx": DeepXBackend,
         "qnn": QNNBackend,
         "litert": LiteRTBackend,
+        "hailo": HailoBackend,
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/autobackend.py:165]: 💡 **MEDIUM**: Direct `.hef` paths still cannot reach this backend. `AutoBackend._model_type()` only recognizes the Hailo export directory suffix (`_hailo_model` from `export_formats()`), so `YOLO("model.hef").predict(...)` resolves to no format and raises the unsupported-format error before `HailoBackend.load_model()` can use its `w.suffix == ".hef"` branch. Add explicit `.hef` format detection (or remove the unreachable file-path branch).

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The backend integration and Hailo output decoding are otherwise coherently wired into AutoBackend. One routing gap prevents the advertised direct HEF-file loading path from working through predict or val.

💬 Posted 1 inline comment

**Comment by glenn-jocher** [comment]
@JESUSROYETH thank you again for the device-level findings and validated YOLO26 decoder contract. Could you please review this implementation and validate `predict` and `val` on your Hailo-8L for YOLOv8/YOLO11 NMS output and YOLO26 raw one-to-one output? GitHub would not accept a formal review request because external users must first be repository collaborators, so I am tagging you here instead.

File: ultralytics/nn/autobackend.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/nn/autobackend.py:165]: Resolved by deleting the unreachable direct-file branch. Hailo inference intentionally loads the exported `_hailo_model` directory so `metadata.yaml` remains available for class names and model-family post-processing.

**Comment by JESUSROYETH** [comment]
Hi @glenn-jocher 👋

I ran #25247 on the Hailo-8L (PCIe, HailoRT 4.23), with the full `predict`/`val` pipeline and the three export paths. It works, and the numbers are consistent:

| model | `val` mAP50 / mAP50-95 (HEF) | `val` mAP50 / mAP50-95 (PyTorch, same protocol) |
|---|---|---|
| yolo11n (on-chip NMS) | 0.531 / 0.415 | 0.672 / 0.504 |
| yolo26n (host decode) | 0.605 / 0.447 | 0.642 / 0.476 |

While waiting for your reply I had written the same backend locally (I never opened the PR, you were faster), and it turned out useful as a cross-check: your decoder matches my independent implementation to the 4th decimal on both models, and `yolov8n_hailo_model` predict returns the same detections too. About the yolo11n gap vs PyTorch: it comes from the fixed `conf=0.25` inside the NMS exports, so `val` only integrates the curve above it, as the docs now explain. At matched conf the retention is 96.5% (#25231). Sequential model loading in one process also works, and the resource release in 4f0d044 is important: without releasing the device, the second load crashes the driver. I saw the same crash in my local version.

Also, a small note about performance. `forward` re-activates the network group on every frame, and `_decode_raw` rebuilds the anchors on every frame, both are per-model constants. Keeping the group activated for the backend lifetime and caching the anchors gives, on the 8L (100 frames x 2 alternating rounds):

| model | as in the PR | persistent activation + cached anchors |
|---|---|---|
| yolo11n | 30.7 ms/frame | 27.2 ms/frame (−11%) |
| yolo26n | 36.0 ms/frame | 30.8 ms/frame (−14%) |

The change is small. In `load_model`:

```python
self._activation = network_group.create_params()
stack.enter_context(network_group.activate(self._activation))  # keep activated for the backend lifetime
```

then `forward` calls `self.model.infer(...)` without the `with` block, and `_decode_raw` caches the `make_anchors` result on first use. mAP is unchanged after both changes (re-validated: 0.5308/0.4146 and 0.6050/0.4470, identical). I'm happy to push the exact diff I benchmarked if you think it's worth it .. let me know.

**Comment by glenn-jocher** [comment]
@JESUSROYETH oh nice improvement! Yes please submit a PR targeting this branch if you can.

**Comment by JESUSROYETH** [comment]
Hi @glenn-jocher 👋

I kept investigating why YOLO26 loses more accuracy than YOLOv8/YOLO11 when quantised for Hailo (93.4% mAP50 retention vs 96.5-100% with the same default calibration). I ran a controlled matrix on the 8L: same checkpoint, same 128 calibration images, same QAT settings, only one variable changed at a time.

| yolo26n variant | mAP50 retention vs its PyTorch |
|---|---|
| e2e head (one-to-one), QAT | 93.4% |
| one-to-many head + NMS, QAT | 91.5% |
| e2e head, no QAT | 85.7% |
| e2e head, QAT with 1238 out-of-domain images | 85.7% |

Three conclusions from the numbers:

1. The gap is not in the NMS-free head. The one-to-many head of the same checkpoint degrades the same or worse, so the cause is in the body — most likely the attention path. I could not isolate it further: the DFC only supports 8-bit inputs for `matmul` layers on Hailo-8/8L (`a8_w8*` modes in the toolchain), and the allocator also rejects 16-bit matmul outputs for this graph, so there is no option to run the attention products at higher precision on this hardware generation. I also tried the official Hailo model-zoo recipe for yolo26s (16-bit on the depthwise layers and on the output convs): it compiles fine but the mAP does not move (0.426 vs 0.427), so the noise looks spread across the int8 body rather than concentrated in one layer you could protect.

2. QAT must be in-domain. Without fine-tuning the retention drops to 85.7%, and fine-tuning with 1238 out-of-domain images gives the same 85.7% — out-of-domain QAT is equivalent to no QAT. This explains the "more images made it worse" result I reported earlier.

3. Half of the deployment gap is recoverable. The HEF scores come out shifted down ~0.05 on average, so `conf=0.20` on the HEF matches the detection count of PyTorch at 0.25, and `conf=0.15` matches its mAP50. The other half is real ranking noise (20.8% of intra-image score pairs invert), no threshold recovers that.

The practical summary, if useful for the docs: deploy YOLO26 HEFs with `conf` about 0.05 lower than tuned in PyTorch, always calibrate in-domain, and expect ~93-94% retention on Hailo-8L against ~96.5% for YOLO11 — when accuracy is the priority and the model is swappable, YOLO11 quantises better on this hardware generation.

**Comment by UltralyticsAssistant** [comment]
🎉 Merged in #25247—thank you @glenn-jocher, with valuable contributions from @JESUSROYETH!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This work brings native HailoRT inference to Ultralytics, enabling streamlined `predict` and `val` workflows for YOLOv8, YOLO11, and YOLO26 on Hailo hardware. The new backend, automatic output decoding, and improved documentation make deployment simpler while preserving compatibility with existing Hailo pipelines. Thank you for helping make edge AI more accessible and practical!

**Comment by glenn-jocher** [comment]
@JESUSROYETH just saw your last post, seems super useful for docs, I'll add in a new PR.

**Comment by glenn-jocher** [comment]
Or actually can you do the PR? Then you can show up in the Hailo docs in the authors section :)
